File: Objective-C_IBDesignables&IBInspectables/Objective-C_IBDesignables&IBInspectables/Classes/PieChartView.m
//
//  PieChartView.m
//  Objective-C_IBDesignables&IBInspectables
//
//  Created by Kaê Angeli Coutinho on 2/26/15.
//  Copyright (c) 2015 Kaê Angeli Coutinho. All rights reserved.
//

#import "PieChartView.h"

@interface PieChartView()

@end

@implementation PieChartView

@synthesize pieBackgroundLayer;
@synthesize pieLayer;
@synthesize pieStrokeWidth;
@synthesize piePercentage;
@synthesize pieBackgroundColor;
@synthesize pieColor;

#pragma mark - Constructors

-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if(self)
    {
        [self setPieBackgroundLayer:[CAShapeLayer new]];
        [self setPieLayer:[CAShapeLayer new]];
        [self setPieStrokeWidth:0.0];
        [self setPiePercentage:0.0];
        [self setPieBackgroundColor:[UIColor colorWithWhite:0.8 alpha:1.0]];
        [self setPieColor:[UIColor orangeColor]];
    }
    return self;
}

-(id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if(self)
    {
        [self setPieBackgroundLayer:[CAShapeLayer new]];
        [self setPieLayer:[CAShapeLayer new]];
        [self setPieStrokeWidth:0.0];
        [self setPiePercentage:0.0];
        [self setPieBackgroundColor:[UIColor colorWithWhite:0.8 alpha:1.0]];
        [self setPieColor:[UIColor orangeColor]];
    }
    return self;
}

#pragma mark - Methods

-(void)layoutSubviews
{
    [super layoutSubviews];
    [self setPieStrokeWidth:[self frame].size.width / 2.0];
    if([self pieBackgroundLayer] != nil)
    {
        [self setPieBackgroundLayer:[CAShapeLayer new]];
        [[self layer] addSublayer:[self pieBackgroundLayer]];
        CGRect circleRect = CGRectInset([self bounds],
                                        ([self pieStrokeWidth] / 2.0),
                                        ([self pieStrokeWidth] / 2.0));
        UIBezierPath * circlePath = [UIBezierPath bezierPathWithOvalInRect:circleRect];
        [[self pieBackgroundLayer] setPath:[circlePath CGPath]];
        [[self pieBackgroundLayer] setFillColor:nil];
        [[self pieBackgroundLayer] setLineWidth:[self pieStrokeWidth]];
        [[self pieBackgroundLayer] setStrokeColor:[[self pieBackgroundColor] CGColor]];
    }
    [[self pieBackgroundLayer] setFrame:[[self layer] bounds]];
    if([self pieLayer] != nil)
    {
        [self setPieLayer:[CAShapeLayer new]];
        [[self layer] addSublayer:[self pieLayer]];
        CGRect circleRect = CGRectInset([self bounds],
                                        ([self pieStrokeWidth] / 2.0),
                                        ([self pieStrokeWidth] / 2.0));
        UIBezierPath * circlePath = [UIBezierPath bezierPathWithOvalInRect:circleRect];
        [[self pieLayer] setPath:[circlePath CGPath]];
        [[self pieLayer] setFillColor:nil];
        [[self pieLayer] setLineWidth:[self pieStrokeWidth]];
        [[self pieLayer] setStrokeColor:[[self pieColor] CGColor]];
        
        [[self pieLayer] setAnchorPoint:CGPointMake(0.5,0.5)];
        [[self pieLayer] setTransform:CATransform3DRotate([[self pieLayer] transform],(-M_PI / 2.0),0.0,0.0,1.0)];
    }
    [[self pieLayer] setFrame:[[self layer] bounds]];
    [self updateLayerProperties];
}

-(void)updateLayerProperties
{
    if([self pieLayer] != nil)
    {
        [[self pieLayer] setStrokeEnd:([self piePercentage] / 100.0)];
    }
}

-(void)updatePiePercentage:(double)newPiePercentage
{
    if([self pieLayer] != nil)
    {
        [CATransaction begin];
        CABasicAnimation * animation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
        [animation setDuration:((newPiePercentage / 100.0) - ([self piePercentage] / 100.0)) * 3.0];
        [animation setFromValue:[NSNumber numberWithDouble:[self piePercentage] / 100.0]];
        [animation setToValue:[NSNumber numberWithDouble:(newPiePercentage / 100.0)]];
        [animation setTimingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut]];
        [CATransaction setCompletionBlock:^
        {
            [CATransaction begin];
            [CATransaction setValue:kCFBooleanTrue forKey:kCATransactionDisableActions];
            [[self pieLayer] setStrokeEnd:(newPiePercentage / 100.0)];
            [CATransaction commit];
        }];
        [[self pieLayer] addAnimation:animation forKey:@"animateStrokeEnd"];
        [CATransaction commit];
    }
}

@end

File: Swift_IBDesignables&IBInspectables/Swift_IBDesignables&IBInspectables/Classes/PieChartView.swift
//
//  PieChart.swift
//  Swift_IBDesignables&IBInspectables
//
//  Created by Kaê Angeli Coutinho on 2/26/15.
//  Copyright (c) 2014 Kaê Angeli Coutinho. All rights reserved.
//

import UIKit
import QuartzCore

@IBDesignable
class PieChartView: UIView
{
    // MARK: Properties
    
    var pieBackgroundLayer: CAShapeLayer!
    var pieLayer: CAShapeLayer!
    var pieStrokeWidth: Double = 0.0
    
    @IBInspectable
    var piePercentage: Double = 0.0
    
    @IBInspectable
    var pieBackgroundColor: UIColor = UIColor(white:0.8,alpha:1.0)
    
    @IBInspectable
    var pieColor: UIColor = UIColor.orangeColor()
    
    // MARK: Methods
    
    override func layoutSubviews()
    {
        super.layoutSubviews()
        self.pieStrokeWidth = Double(self.frame.size.width) / 2.0
        if !(self.pieBackgroundLayer != nil)
        {
            self.pieBackgroundLayer = CAShapeLayer()
            self.layer.addSublayer(self.pieBackgroundLayer)
            let circleRect = CGRectInset(self.bounds,
                                         CGFloat(self.pieStrokeWidth / 2.0),
                                         CGFloat(self.pieStrokeWidth / 2.0))
            let circlePath = UIBezierPath(ovalInRect:circleRect)
            self.pieBackgroundLayer.path = circlePath.CGPath
            self.pieBackgroundLayer.fillColor = nil
            self.pieBackgroundLayer.lineWidth = CGFloat(self.pieStrokeWidth)
            self.pieBackgroundLayer.strokeColor = self.pieBackgroundColor.CGColor
        }
        self.pieBackgroundLayer.frame = self.layer.bounds
        if !(self.pieLayer != nil)
        {
            self.pieLayer = CAShapeLayer()
            self.layer.addSublayer(self.pieLayer);
            let circleRect = CGRectInset(self.bounds,
                                         CGFloat(self.pieStrokeWidth / 2.0),
                                         CGFloat(self.pieStrokeWidth / 2.0))
            let circlePath = UIBezierPath(ovalInRect:circleRect)
            self.pieLayer.path = circlePath.CGPath
            self.pieLayer.fillColor = nil
            self.pieLayer.lineWidth = CGFloat(self.pieStrokeWidth)
            self.pieLayer.strokeColor = pieColor.CGColor
            self.pieLayer.anchorPoint = CGPointMake(0.5,0.5)
            self.pieLayer.transform = CATransform3DRotate(self.pieLayer.transform,CGFloat(-M_PI / 2.0),0.0,0.0,1.0)
        }
        self.pieLayer.frame = self.layer.bounds
        self.updateLayerProperties()
    }
    
    func updateLayerProperties()
    {
        if (self.pieLayer != nil)
        {
            self.pieLayer.strokeEnd = CGFloat(self.piePercentage / 100.0)
        }
    }
    
    func updatePiePercentage(#newPiePercentage: Double)
    {
        if (self.pieLayer != nil)
        {
            CATransaction.begin()
            var animation = CABasicAnimation(keyPath:"strokeEnd")
            animation.duration = ((newPiePercentage / 100.0) - (self.piePercentage / 100.0)) * 3.0
            animation.fromValue = self.piePercentage / 100.0
            animation.toValue = newPiePercentage / 100.0
            animation.timingFunction = CAMediaTimingFunction(name:kCAMediaTimingFunctionEaseInEaseOut)
            CATransaction.setCompletionBlock(
            { () -> Void in
                CATransaction.begin()
                CATransaction.setValue(kCFBooleanTrue,forKey:kCATransactionDisableActions)
                self.pieLayer.strokeEnd = CGFloat(newPiePercentage / 100.0)
                CATransaction.commit()
            })
            self.pieLayer.addAnimation(animation,forKey:"animateStrokeEnd")
            CATransaction.commit()
        }
    }
}